framework/base/StaticInstanceTrait.php

```diff
@@ -0,0 +1,41 @@
+<?php
+/**
+ * @link http://www.yiiframework.com/
+ * @copyright Copyright (c) 2008 Yii Software LLC
+ * @license http://www.yiiframework.com/license/
+ */
+
+namespace yii\base;
+
+use Yii;
+
+/**
+ * StaticInstanceTrait provides methods to satisfy [[StaticInstanceInterface]] interface.
+ *
+ * @see StaticInstanceInterface
+ *
+ * @author Paul Klimov <klimov.paul@gmail.com>
+ * @since 2.0.13
+ */
+trait StaticInstanceTrait
+{
+    /**
+     * @var static[] static instances in format: `[className => object]`
+     */
+    private static $_instances = [];
+
+
+    /**
+     * Returns static class instance, which can be used to obtain meta information.
+     * @param bool $refresh whether to re-create static instance even, if it is already cached.
+     * @return static class instance.
+     */
+    public static function instance($refresh = false)
```

Comment:
I know almost nobody will review it, but really it is a singleton guys in 2017. I mean ok, but just rename it to actually match some naming convention? or rename interface? `singletoneInstance` or whatever. As for me this interface just mixing two things - some simple identity map (from how it is used in AR) and bad naming convention.

Reply:
It is not exactly a singleton. Name matches what's there well: trait enables getting a static instance but, unlike singleton, doesn't prevent using constructor.

Map is needed because else a single instance will be shared by classes using this trait no matter what the class is.

Reply:
I think we need a separate chapter in a docs 
witch details about this new functional:
- difference between StaticInstance and Singelton pattern
- Usage case (with examples i.e.)
- Best practics

Reply:
Yes, could be useful. If you have ideas on where to mention all that, feel free to create a pull request. Perfect language isn't necessary, we'll edit it together.